File: HueQuickStartApps/HueQuickStartApp-iOS/HueQuickStartApp/Bridge/PHBridgeController.m
/*******************************************************************************
 Copyright (C) 2017 Philips Lighting Holding B.V.
 All Rights Reserved.
 ********************************************************************************/

#import "PHBridgeController.h"

@interface PHBridgeController ()

@property (nonatomic, weak) PHRootViewController* rootViewController;

@property (nonatomic, strong, readwrite) PHSBridge* bridge;
@property (nonatomic, assign) PHSBridgeConnectionEvent lastConnectionEvent;
@property (nonatomic, assign) BOOL canConnect;

@property (nonatomic, assign) BOOL loadingWheelActive;
@property (nonatomic, assign) BOOL alertActive;
@property (nonatomic, assign) BOOL pushlinkActive;

@end

@implementation PHBridgeController

- (instancetype)initWithBridgeInfo:(PHBridgeInfo *)bridgeInfo andRootViewController:(PHRootViewController *)rootViewController andDelegate:(id<PHBridgeControllerDelegate>)delegate {
    self = [super init];
    
    if (self) {
        self.rootViewController = rootViewController;
        
        self.bridge = [self buildBridgeWithInfo:bridgeInfo];
        self.lastConnectionEvent = PHSBridgeConnectionEventNone;
        self.canConnect = YES;
        
        self.loadingWheelActive = NO;
        self.alertActive = NO;
        self.pushlinkActive = NO;
        
        self.delegate = delegate;
    }
    
    return self;
}

- (void)connect {
    if (self.canConnect) {
        self.canConnect = NO;
        
        [self showConnecting];
    
        [self.bridge connect];
    }
}

- (void)disconnect {
    [self.bridge disconnect];
}

#pragma mark - UI

- (void)showConnecting {
    [self dismissAll];
    
    [self.rootViewController showLoadingWheelWithText:NSLocalizedString(@"Connecting", @"Connecting title on loading screen")];
    
    self.loadingWheelActive = YES;
}

- (void)showCouldNotConnect {
    [self dismissAll];
    
    [self.rootViewController presentViewController:[self createCouldNotConnectAlert]];
    
    self.alertActive = YES;
}

- (UIAlertController *)createCouldNotConnectAlert {
    UIAlertController* couldNotConnectAlert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Could not connect", @"Title alert when couldn't connect to bridge")
                                                                                 message:@""
                                                                          preferredStyle:UIAlertControllerStyleAlert];
    
    [couldNotConnectAlert addAction:[self createAlertActionForRetry]];
    [couldNotConnectAlert addAction:[self createAlertActionForCancel]];
    
    return couldNotConnectAlert;
}

- (UIAlertAction*)createAlertActionForRetry{
    __weak typeof(self) weakSelf = self;
    
    return [UIAlertAction actionWithTitle:NSLocalizedString(@"Retry", @"Retry button on could not connect alert")
                                    style:UIAlertActionStyleDefault
                                  handler:^(UIAlertAction* action) {
                                      weakSelf.alertActive = NO;
                                      
                                      [weakSelf connect];
                                  }];
}

- (UIAlertAction*)createAlertActionForCancel{
    __weak typeof(self) weakSelf = self;
    
    return [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel button on could not connect alert")
                                    style:UIAlertActionStyleDefault
                                  handler:^(UIAlertAction* action) {
                                      weakSelf.alertActive = NO;
                                      
                                      [weakSelf.delegate disconnected];
                                  }];
}

- (void)showConnectionLost {
    [self dismissAll];
    
    [self.rootViewController showLoadingWheelWithText:NSLocalizedString(@"Reconnecting", @"Reconnecting title on loading screen")];
    
    self.loadingWheelActive = YES;
}

- (void)showPushlinkRequired {
    [self dismissAll];
    
    PHBridgePushLinkViewController* pushlinkViewController = [[PHBridgePushLinkViewController alloc] initWithRootViewController:self.rootViewController andDelegate:self];
    
    [self.rootViewController pushViewController:pushlinkViewController];
    
    self.pushlinkActive = YES;
}

- (void)dismissLoadingWheel {
    if (self.loadingWheelActive) {
        [self.rootViewController hideLoadingWheel];
        
        self.loadingWheelActive = NO;
    }
}

- (void)dismissAlert {
    if (self.alertActive) {
        [self.rootViewController dismissViewController];
        
        self.alertActive = NO;
    }
}

- (void)dismissPushlink {
    if (self.pushlinkActive) {
        [self.rootViewController popViewController];
        
        self.pushlinkActive = NO;
    }
}

- (void)dismissAll {
    [self dismissLoadingWheel];
    [self dismissAlert];
    [self dismissPushlink];
}

#pragma mark - PHSBridge

- (PHSBridge *)buildBridgeWithInfo:(PHBridgeInfo *)bridgeInfo {
    // Replace app name and device name
    return [PHSBridge bridgeWithBlock:^(PHSBridgeBuilder* builder) {
        builder.connectionTypes = PHSBridgeConnectionTypeLocal;
        builder.ipAddress = bridgeInfo.ipAddress;
        builder.bridgeID  = bridgeInfo.uniqueId;
        
        builder.bridgeConnectionObserver = self;
        
        [builder addStateUpdateObserver:self];
    } withAppName:@"MyApp" withDeviceName:@"MyDevice"];
}

#pragma mark - PHSBridgeConnectionObserver 

- (void)bridgeConnection:(PHSBridgeConnection *)bridgeConnection handleEvent:(PHSBridgeConnectionEvent)connectionEvent {
    switch (connectionEvent) {
        case PHSBridgeConnectionEventCouldNotConnect:
            [self handleCouldNotConnect];
            break;
        case PHSBridgeConnectionEventConnected:
            // Inform delegate only when also authenticated
            break;
        case PHSBridgeConnectionEventConnectionLost:
            [self handleConnectionLostEvent];
            break;
        case PHSBridgeConnectionEventConnectionRestored:
            [self handleConnectionRestoredEvent];
            break;
        case PHSBridgeConnectionEventDisconnected:
            [self handleDisconnectedEvent];
            break;
            
        case PHSBridgeConnectionEventNotAuthenticated:
        case PHSBridgeConnectionEventLinkButtonNotPressed:
            [self handlePushlinkEvent];
            break;
        case PHSBridgeConnectionEventAuthenticated:
            [self handleAuthenticatedEvent];
            break;
            
        default:
            break;
    }
    
    self.lastConnectionEvent = connectionEvent;
}

- (void)handleCouldNotConnect {
    self.canConnect = YES;
    
    [self showCouldNotConnect];
}

- (void)handleConnectionLostEvent {
    [self showConnectionLost];
}

- (void)handleConnectionRestoredEvent {
    [self dismissAll];
}

- (void)handleDisconnectedEvent {
    self.canConnect = YES;
    
    [self.delegate disconnected];
}

- (void)handlePushlinkEvent {
    if (self.lastConnectionEvent != PHSBridgeConnectionEventNotAuthenticated && self.lastConnectionEvent != PHSBridgeConnectionEventLinkButtonNotPressed) {
        [self showPushlinkRequired];
    }
}

- (void)handleAuthenticatedEvent {
    [self dismissAll];
}

- (void)bridgeConnection:(PHSBridgeConnection *)bridgeConnection handleErrors:(NSArray<PHSError *> *)connectionErrors { }

#pragma mark - PHSBridgeStateUpdateObserver

- (void)bridge:(PHSBridge *)bridge handleEvent:(PHSBridgeStateUpdatedEvent)updateEvent {
    if (updateEvent == PHSBridgeStateUpdatedEventInitialized) {
        [self.delegate connected];
        
        [self startHeartbeat];
    }
}

- (void)startHeartbeat {
    PHSBridgeConnection* connection = self.bridge.bridgeConnections.firstObject;
    
    [connection.heartbeatManager startHeartbeatWithType:PHSBridgeStateCacheTypeFullConfig interval:10];
}

#pragma mark - PHBridgePushLinkViewControllerDelegate

- (void)pushlinkCancelled {
    [self dismissAll];
    
    [self disconnect];
}

@end

File: HueQuickStartApps/HueQuickStartApp-iOS/HueQuickStartAppSwift/Bridge/PHBridgeController.swift
/*******************************************************************************
 Copyright (C) 2017 Philips Lighting Holding B.V.
 All Rights Reserved.
 ********************************************************************************/

import Foundation
import UIKit

protocol PHBridgeControllerDelegate {
    func connected()
    func disconnected()
}

class PHBridgeController : NSObject {
    let alertHandler:AlertPresenting?
    let pushLinkHandler:PushLinkPresenting?
    let activityIndicator:ActivityDisplaying?
    var delegate:PHBridgeControllerDelegate?
    
    var lastConnectionEvent:PHSBridgeConnectionEvent
    var canConnect:Bool
    
    let bridgeInfo:PHBridgeInfo
    lazy var bridge:PHSBridge = {
        return buildBridge(info:bridgeInfo)
    }()
    
    init?(bridgeInfo:PHBridgeInfo?, activityIndicator:ActivityDisplaying?, alertHandler:AlertPresenting?, pushLinkHandler:PushLinkPresenting?, delegate:PHBridgeControllerDelegate) {
        guard let bridgeInfo = bridgeInfo else {
            return nil
        }
        
        self.bridgeInfo = bridgeInfo
        self.alertHandler = alertHandler
        self.pushLinkHandler = pushLinkHandler
        self.activityIndicator = activityIndicator
        self.delegate = delegate
        
        self.canConnect = true
        self.lastConnectionEvent = .none
        
        super.init()
    }
    
    func buildBridge(info:PHBridgeInfo) -> PHSBridge {
        // Replace app name and device name
        return PHSBridge.init(block: { (builder) in
            builder?.connectionTypes = .local
            builder?.ipAddress = info.ipAddress
            builder?.bridgeID  = info.uniqueId
            
            builder?.bridgeConnectionObserver = self
            builder?.add(self)
        }, withAppName: "MyApp", withDeviceName: "MyDevice")
    }
    
    func connect () {
        if canConnect {
            showConnecting()
            bridge.connect()
        }
    }
    
    func disconnect() {
        bridge.disconnect()
    }
    
    func startHeartBeat() {
        if let connection:PHSBridgeConnection = self.bridge.bridgeConnections().first {
            connection.heartbeatManager.startHeartbeat(with:.fullConfig, interval: 10)
        }
    }
    
    func dismissAll() {
        alertHandler?.dismissAlerts()
        pushLinkHandler?.dismissPushLinkScreen()
        activityIndicator?.showLoadingWheel(show: false, message: "")
    }
}

private typealias PHBridgeControllerAlerts = PHBridgeController
extension PHBridgeControllerAlerts {
    func showCouldNotConnectAlert() {
        if let alertHandler = alertHandler {
            alertHandler.presentAlert(alert:couldNotConnectAlert())
        }
    }
    
    func couldNotConnectAlert() -> UIAlertController {
        let couldNotConnectAlert:UIAlertController = UIAlertController(title: NSLocalizedString("Could not connect", comment:"Title alert when couldn't connect to bridge"),
                                                                       message: nil,
                                                                       preferredStyle:.alert)
        couldNotConnectAlert.addAction(alertActionForRetry())
        couldNotConnectAlert.addAction(alertActionForCancel())
        
        return couldNotConnectAlert
    }
    
    func alertActionForRetry() -> UIAlertAction {
        return UIAlertAction(title: NSLocalizedString("Retry", comment:"Retry button on could not connect alert"),
                             style:.default,
                             handler: { [weak self] (action) in
                                if let strongSelf = self {
                                    strongSelf.connect()
                                }
        })
    }
    
    func alertActionForCancel() -> UIAlertAction {
        return UIAlertAction(title: NSLocalizedString("Cancel", comment:"Cancel button on could not connect alert"),
                             style:.default,
                             handler: { [weak self] (action) in
                                if let strongSelf = self {
                                    strongSelf.delegate?.disconnected()
                                }
        })
    }
}

private typealias PHBridgeControllerActivityIndicator = PHBridgeController
extension PHBridgeControllerActivityIndicator {
    func showConnecting() {
        if let activityIndicator = activityIndicator {
            activityIndicator.showLoadingWheel(show: true, message:NSLocalizedString("Connecting", comment:"Connecting title on loading screen"))
        }
    }
    
    func showConnectionLost() {
        if let activityIndicator = activityIndicator {
            activityIndicator.showLoadingWheel(show: true, message:NSLocalizedString("Reconnecting", comment:"Reconnecting title on loading screen"))
        }
    }
}

private typealias PHBridgeControllerEvents = PHBridgeController
extension PHBridgeControllerEvents {
    func handleConnectionLostEvent() {
        showConnectionLost()
    }
    
    func handleConnectionRestoredEvent() {
        dismissAll()
    }
    
    func handleDisconnectedEvent() {
        self.canConnect = true
        
        // Disconnected event is sent directly after PHSBridgeConnectionEventCouldNotConnect, but user should decide to retry before informing delegate
        if lastConnectionEvent != .couldNotConnect {
            delegate?.disconnected()
        }
    }
    
    func handlePushLinkEvent() {
        if lastConnectionEvent != .notAuthenticated && lastConnectionEvent != .linkButtonNotPressed {
            self.dismissAll()
            pushLinkHandler?.presentPushLinkScreen(completion: { [weak self] (success) in
                if !success {
                    if let strongSelf = self {
                        strongSelf.bridge.disconnect()
                    }
                }
            })
        }
    }
    
    func handleAuthenticationEvent() {
        dismissAll()
    }
}

private typealias PHBridgeConnectionObserverConformance = PHBridgeController
extension PHBridgeConnectionObserverConformance : PHSBridgeConnectionObserver {
    func bridgeConnection(_ bridgeConnection: PHSBridgeConnection!, handle connectionEvent: PHSBridgeConnectionEvent) {
        switch connectionEvent {
        case .couldNotConnect:
            showCouldNotConnectAlert()
        case .connected:
            // Inform delegate only when also authenticated
            break
        case .connectionLost:
            handleConnectionLostEvent()
        case .connectionRestored:
            handleConnectionRestoredEvent()
        case .disconnected:
            handleDisconnectedEvent()
        case .notAuthenticated:
            fallthrough
        case .linkButtonNotPressed:
            handlePushLinkEvent()
        case .authenticated:
            handleAuthenticationEvent()
        default:
            break
        }
        
        self.lastConnectionEvent = connectionEvent;
    }
    
    func bridgeConnection(_ bridgeConnection: PHSBridgeConnection!, handleErrors connectionErrors: [PHSError]!) {
        
    }
}

private typealias PHBridgeStateUpdateObserverConformance = PHBridgeController
extension PHBridgeStateUpdateObserverConformance : PHSBridgeStateUpdateObserver {
    func bridge(_ bridge: PHSBridge!, handle updateEvent: PHSBridgeStateUpdatedEvent) {
        if updateEvent == .initialized {
            self.delegate?.connected()
            startHeartBeat()
        }
    }
}
